chain-core/src/state/tendermint.rs

```diff
@@ -0,0 +1,129 @@
+use parity_codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+use std::{fmt, ops};
+
+/// Tendermint block height
+/// TODO: u64?
+pub type BlockHeight = i64;
+
+/// The protobuf structure currently has "String" to denote the type / length
+/// and variable length byte array. In this internal representation,
+/// it's desirable to keep it restricted and compact. (TM should be encoding using the compressed form.)
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize, Encode, Decode)]
+pub enum TendermintValidatorPubKey {
+    Ed25519([u8; 32]),
+    // there's PubKeySecp256k1, but https://tendermint.com/docs/spec/abci/apps.html#validator-updates
+    // "The pub_key currently supports only one type:"
+    // "type = "ed25519" anddata = <raw 32-byte public key>`"
+    // there's also PubKeyMultisigThreshold, but that probably wouldn't be used for individual nodes / validators
+    // TODO: some other schemes when they are added in TM?
+}
+
+impl TendermintValidatorPubKey {
+    pub fn to_validator_update(&self) -> (String, Vec<u8>) {
+        match self {
+            TendermintValidatorPubKey::Ed25519(key) => {
+                let mut v = Vec::with_capacity(32);
+                v.extend_from_slice(&key[..]);
+                ("ed25519".to_string(), v)
+            }
+        }
+    }
+}
+
+/// "Note that the maximum total power of the validator set is bounded by MaxTotalVotingPower = MaxInt64 / 8.
+/// Applications are responsible for ensuring they do not make changes to the validator set that cause it to exceed this limit."
+const TENDERMINT_MAX_VOTE_POWER: i64 = std::i64::MAX / 8;
+
+/// Tendermint consensus voting power
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub struct TendermintVotePower(i64);
```

Comment:
Why are we using `i64` here? Can voting power ever be negative?

Reply:
Because that's how it is in Tendermint protobuf refs; so, one can do `.into()`.
It shouldn't ever be negative, but protobuf defs won't reflect that. we may perhaps move this wrapping definitions to rust-abci later